components/locale_canonicalizer/src/locale_canonicalizer.rs

```diff
@@ -14,9 +18,78 @@ pub enum CanonicalizationResult {
 }
 
 pub struct LocaleCanonicalizer<'a> {
+    aliases: DataPayload<'a, AliasesV1>,
     likely_subtags: DataPayload<'a, LikelySubtagsV1>,
 }
 
+#[inline]
+fn uts35_rule_matches(source: &Locale, ruletype: &LanguageIdentifier) -> bool {
+    (ruletype.language.is_empty() || ruletype.language == source.id.language)
+        && (ruletype.script.is_none() || ruletype.script == source.id.script)
+        && (ruletype.region.is_none() || ruletype.region == source.id.region)
+        && (ruletype.variants.is_empty()
+            || ruletype
+                .variants
+                .iter()
+                .all(|v| source.id.variants.contains(v)))
+}
+
+fn uts35_replacement(
+    source: &mut Locale,
+    ruletype_has_language: bool,
+    ruletype_has_script: bool,
+    ruletype_has_region: bool,
+    ruletype_variants: Option<&subtags::Variants>,
+    replacement: &LanguageIdentifier,
+) {
+    if ruletype_has_language || source.id.language.is_empty() && !replacement.language.is_empty() {
```

Comment:
`suggestion (non-blocking)`: would it make sense to assume that if we have `ruletype_has_language` then `replacement.language` is not empty?

Reply:
Yes, this is a bit of leftover code, I used to use this for scripts and regions as well, and now this only used for language rules. I can simplify things here a bit.

Reply:
Hmm, well I hit some test failures with that, so I think I'd prefer to leave this as-is. The intention is that there are two separate cases where a replacement is made:
```
A matching rule can be used to transform the source fields as follows

    if type.field ≠ {}
        source.field = (source.field - type.field) ∪ replacement.field
    else if source.field = {} and replacement.field ≠ {}
        source.field = replacement.field
```
I'll add parentheses to make this intention clearer.